vision/module/location.py

```diff
@@ -297,38 +297,57 @@ def _increase_brightness(self, offset: int) -> np.ndarray:
 
-        # Convert BoundingBoxes to list of values per axis
-        vertices = np.array(
-            [vertex for bbox in self.text_boxes for vertex in bbox.vertices]
-        )  # unravel into list of vertices
-        axes_vals = vertices.transpose()  # axis_vals[0] = x, axis_vals[1] = y
-
-        # Find max and min values for each axis
-        x_ubound = np.max(axes_vals[0])  # horizontal upper bound of text
-        x_lbound = np.min(axes_vals[0])  # horizontal lower bound of text
-        y_lbound = np.max(axes_vals[1])  # vertical lower bound of text
-
         # Filter out rectangles that are not directly below text
         circle_filter = np.array(
             [], dtype=np.uint16
         )  # array of indices of circles out of bounds
 
+        # get text bounding extreme points of rotated bounding boxes
+        ll_x, ll_y = self.text_boxes[0].vertices[0]  # lower left point
+        # ul_x, ul_y = self.text_boxes[0].vertices[1] #NOTE: not needed
+        # ur_x, ur_y = self.text_boxes[-1].vertices[2] #NOTE: not needed
+        lr_x, lr_y = self.text_boxes[-1].vertices[3]  # lower right point
+
+        rot_neg90 = np.array([[0, 1], [-1, 0]])  # matrix to rotate a vector -90 degrees
+
+        # bounding vectors housing the allowed region
+        lr_vect = (lr_x - ll_x, lr_y - ll_y)  # vector from ll bound to lr bound of text
+        ll_vect = (lr_x - ll_x, lr_y - ll_y)  # vector from lr bound to ll bound of text
+
+        # horizonal bounding vectors perpendicular to lower bound vectors (pointed downwards)
+        left_vect = tuple(np.dot(lr_vect, rot_neg90).astype(int))  # left bound vector
+        right_vect = tuple(np.dot(ll_vect, rot_neg90).astype(int))  # right bound vector
+
+        # create a vector from the lower-left/lower-right corner of text to all detected circles
+        # the cross product tells us what side of the bounding vectors the circle is on
         for i, (x, y, _) in enumerate(self.circles):
-            if y <= y_lbound:  # remove if above lower bound of text
-                circle_filter = np.append(circle_filter, i)
-            elif (
-                x <= x_lbound or x >= x_ubound
-            ):  # remove if horizontally outside text bounds
+            needs_removal = False  # assume the circle is safely below the text
+
+            # create two vectors
+            ll_to_circ = (x - ll_x, y - ll_y)  # vector from ll bound to circle
+            lr_to_circ = (x - lr_x, y - lr_y)  # vector from lr bound to circle
+
+            lower_cross = np.cross(lr_vect, ll_to_circ)  # <0 = above lower bound
+            if lower_cross <= 0:
+                needs_removal = True
+            left_cross = np.cross(left_vect, ll_to_circ)  # >0 = outside left bound
+            if left_cross >= 0:
+                needs_removal = True
+            right_cross = np.cross(right_vect, lr_to_circ)  # <0 = outside right bound
+            if right_cross <= 0:
+                needs_removal = True
```

Comment:
Really good idea taking the cross products to determine location.

Reply:
Thanks